src/Compatibility/Microsoft.DotNet.PackageValidation/Package.cs

```diff
@@ -0,0 +1,178 @@
+            {
+                if (string.IsNullOrEmpty(value))
+                {
+                    throw new ArgumentException($"{value} is empty or null.  Please check the package path.");
+                }
+
+                if (!File.Exists(Path.GetFullPath(value)))
+                {
+                    throw new FileNotFoundException($"{value} doesnt exist. Please check the package path.");
+                }
+                _packagePath = value;
+            }
+        }
+
+        public MemoryStream PackageStream { get; set; }
+
+        public Dictionary<NuGetFramework, IEnumerable<PackageDependency>> PackageDependencies { get; set; }
+
+        public IEnumerable<ContentItem> CompileAssets { get; private set; }
+
+        public IEnumerable<ContentItem> RefAssets { get; private set; }
+
+        public IEnumerable<ContentItem> LibAssets { get; private set; }
+        
+        public IEnumerable<ContentItem> PackageAssets { get; private set; }
+
+        public IEnumerable<ContentItem> RuntimeSpecificAssets { get; private set; }
+
+        public IEnumerable<ContentItem> RuntimeAssets { get; private set; }
+
+        public bool HasRefAssemblies => RefAssets.Any();
+
+        public IEnumerable<string> Rids { get; private set; }
+
+        public IEnumerable<NuGetFramework> FrameworksInPackage { get; private set; }
+
+        public ContentItem FindBestRuntimeAssetForFramework(NuGetFramework framework)
+        {
+            SelectionCriteria managedCriteria = _conventions.Criteria.ForFramework(framework);
+            return _packageAssets.FindBestItemGroup(managedCriteria,
+                _conventions.Patterns.RuntimeAssemblies)?.Items.FirstOrDefault();
+        }
+
+        public ContentItem FindBestRuntimeAssetForFrameworkAndRuntime(NuGetFramework framework, string rid)
+        {
+            SelectionCriteria managedCriteria = _conventions.Criteria.ForFrameworkAndRuntime(framework, rid);
+            return _packageAssets.FindBestItemGroup(managedCriteria,
+                _conventions.Patterns.RuntimeAssemblies)?.Items.FirstOrDefault();
+        }
+
+        public ContentItem FindBestCompileAssetForFramework(NuGetFramework framework)
+        {
+            SelectionCriteria managedCriteria = _conventions.Criteria.ForFramework(framework);
+            if (RefAssets.Any())
+            {
+                return _packageAssets.FindBestItemGroup(managedCriteria,
+                    _conventions.Patterns.CompileRefAssemblies)?.Items.FirstOrDefault(); ;
+            }
+            else
+            {
+                return _packageAssets.FindBestItemGroup(managedCriteria,
+                    _conventions.Patterns.CompileLibAssemblies)?.Items.FirstOrDefault(); ;
+                
+            }
+        }
+
+        private int VersionComparer(int major, int minor, int patch) => major * 100 + minor * 10 + patch;
+
+        public async Task<MemoryStream> DownloadLatestStableVersionAsync(string nugetFeed = _nugetFeed)
```

Comment:
Downloading outside of NuGet's restore invocation is highly problematic. Imagine our 100 packages in dotnet/runtime. For each, we would trigger a NuGet invocation (either in-proc or out-of-proc). In the past we had about 30 NuGet restore invocations in corefx and I worked hard on drilling them down to three:
1. the nuget sdk resolver download (used when specifying an `<Sdk />` element which isn't in the local sdk)
2. the arcade pre-restore (that always happens when using the build.cmd/sh scripts)
3. the repo's restore (that's the NuGet static graph restore that restores all projects in the repo / solution)

With this design, we would add one additional restore invocation per task invocation (which I believe means per package). Every additional NuGet invocation adds risk and results in a non reliable system. In addition to that, restoring late, during the task invocation, instead of early as part of the project's restore adds additional flakiness as it requires the NuGet repository constantly being available (which it isn't in case of hardware of NuGet server issues).

I recommend to use the `Microsoft.DotNet.PackageValidation` project as an msbuild sdk (which it already is) and hook onto one of the APIs that NuGet offers, in this case `CollectPackageDownloads` to add a `PackageDownload` item to the project so that it is downloaded during restore.

I suspect that there's a PackageDownload Version syntax that allows to download the latest available package without the need to retrieve the versions first.

Reply:
@Anipik were you able to think about this? I think addressing these concerns is important, especially for the consumption of this package in dotnet/runtime.

Reply:
Yes i was able to move the download part to the restore phase